gems/aws-sdk-s3/lib/aws-sdk-s3/file_downloader.rb

```diff
@@ -133,32 +107,28 @@ def multithreaded_get_by_parts(parts)
       end
 
       def thread_batches(chunks, param)
-        # create a tmp dir under destination dir for batches
-        dir = Dir.mktmpdir(nil, File.dirname(@path))
-        batches = file_batches(chunks, dir, param)
-        parts = batches.flat_map(&:values)
-        begin
-          batches.each do |batch|
-            threads = []
-            batch.each do |chunk, file|
-              threads << Thread.new do
-                @client.get_object(
-                  :bucket => @bucket,
-                  :key => @key,
-                  param.to_sym => chunk,
-                  :response_target => file
-                )
-              end
+        batches(chunks, param).each do |batch|
+          threads = []
+          batch.each do |chunk|
+            threads << Thread.new do
+              resp = @client.get_object(
+                :bucket => @bucket,
+                :key => @key,
+                param.to_sym => chunk
+              )
+              write(resp)
             end
-            threads.each(&:join)
           end
-          concatenate_parts(parts)
-        ensure
-          # clean up tmp dir
-          FileUtils.remove_entry(dir)
+          threads.each(&:join)
```

Comment:
I wonder if we can squeeze out more performance by going to a thread worker model with thread-safe queues.